### PR TITLE
test/test_socket: fix dispatch_sockets() unexpected exception

### PR DESCRIPTION
**Pull request by rosinL**
In the test test_unexpected_down(), there is client running on CPU0
and server runing on CPU1, the server maybe finished before the client
(client and other workloads share CPU0), so client couldn't read data
and throws `dispatch_sockets(): cb_client() got unexpected 
exception std::system_error (error crimson::net:4, read eof)`

Signed-off-by: luo rixin <luorixin@huawei.com>

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/test/crimson/test_socket.cc [hunk 1]
```diff
@@ -313,7 +313,7 @@ class Connection {
       // successful
       logger.debug("dispatch_read_unbounded(): "
                    "expected error {}", e);
-      shutdown();
+      socket->force_shutdown_in();
```
Comment (cyx1231st) [inline src/test/crimson/test_socket.cc]: Thank you for the fix! This is indeed an issue, but the fix is not appropriate.

The intention of `test_unexpected_down()` is to verify if the server-side can detect an unexpected socket error (from client) and do a clean close properly, with behaviors exactly the same with `crimson::net::SocketConnection`. This means the server-side cannot use `force_shutdown_in()` or `force_shutdown_out()` and should still call `shutdown()`.

IMHO a better fix could be just treat `read-eof` as an expected case and suppress it from client. Likely at L376, change from 
```
    [] (auto cs) { return Connection::dispatch_rw_bounded(cs, 128, true); },
```
to
```
    [] (auto cs) {
      return Connection::dispatch_rw_bounded(cs, 128, true
      ).handle_exception_type([] (const std::system_error& e) {
        logger.debug("test_unexpected_down(): client got error {}", e);
        ceph_assert(e.code() == error::read_eof);
      });
    },
```

Comment (rosinL) [reply src/test/crimson/test_socket.cc]: I see. Is there another expected exception `std::errc::broken_pipe` we should consider?

Comment (cyx1231st) [reply src/test/crimson/test_socket.cc]: I'm not able to reproduce the case of `broken_pipe` by changing to bounded read and unbounded write from client. So currently I don't think this is an expected exception.

But that would be great if you can also reproduce the `broken_pipe` case.

Comment (cyx1231st) [reply src/test/crimson/test_socket.cc]: Could you also fix the typo related to "unexpected" in the PR title, commit title and message?

**Comment by tchaikov** [comment]
@rosinL ping?

**Pull request merged**